Gravity-tube/src/Pages/AuthWidget/SignUp/SignUp.js

```diff
@@ -0,0 +1,389 @@
+      case "SET_CONFIRM_PWD":
+        return { ...state, confirmPassword: payload };
+      default:
+        return state;
+    }
+  };
+
+  const [formState, formDispatch] = useReducer(formReducer, formInitialState);
+
+  const formValidation = () => {
+    let errFlag = true;
+    if (
+      formState.firstName === "" ||
+      !/^[a-zA-Z]+(\s*\w*)*$/.test(formState.firstName)
+    ) {
+      errorsDispatch({
+        type: "SHOW_FIRST_N_ERROR",
+        payload: "Please enter valid name",
+      });
+      errFlag = false;
+    }
+
+    if (
+      formState.lastName === "" ||
+      !/^[a-zA-Z]+(\s*\w*)*$/.test(formState.lastName)
+    ) {
+      errorsDispatch({
+        type: "SHOW_LAST_N_ERROR",
+        payload: "Please enter valid surname",
+      });
+      errFlag = false;
+    }
+
+    if (formState.email === "" || !/^.+@.+\.com$/.test(formState.email)) {
+      errorsDispatch({
+        type: "SHOW_EMAIL_ERROR",
+        payload: "Please enter valid email",
+      });
+      errFlag = false;
+    }
+
+    if (
+      formState.password === "" ||
+      !/^(?=.*\d)(?=.*[a-z])(?=.*[A-Z])(?=.*[a-zA-Z]).{8,}$/g.test(
+        formState.password
+      )
+    ) {
+      errorsDispatch({
+        type: "SHOW_PWD_ERROR",
+        payload:
+          "Password length should contain minimum 8 characters (at least one capital, small letter and number)",
+      });
+      errFlag = false;
+    }
+    if (
+      formState.confirmPassword === "" ||
+      formState.password !== formState.confirmPassword
+    ) {
+      errorsDispatch({
+        type: "SHOW_CONFIRM_PWD_ERROR",
+        payload: "Password does not match",
+      });
+      errFlag = false;
+    }
+    return errFlag;
+  };
+
+  const signupHandler = async (e) => {
+    e.preventDefault();
+    setError("");
+    if (formValidation()) {
+      console.log("hi");
+      try {
+        console.log("try");
+        const userData = await axios.post(`/api/auth/signup`, {
+          firstName: formState.firstName,
+          lastName: formState.lastName,
+          email: formState.email,
+          password: formState.password,
+        });
+        localStorage.setItem("token", userData.data.encodedToken);
+        authDispatch({ type: "SIGN_UP", payload: userData.data.encodedToken });
+        navigate("/");
+        toast.success(' User Sign Up Successfully!!', {
+          position: "top-center",
+          autoClose: 5000,
+          hideProgressBar: false,
+          closeOnClick: true,
+          pauseOnHover: true,
+          draggable: true,
+          progress: undefined,
+          theme: "dark"
+          });
+      } catch (err) {
+        toast.error(' Something went Wrong!!', {
+          position: "top-center",
+          autoClose: 5000,
+          hideProgressBar: false,
+          closeOnClick: true,
+          pauseOnHover: true,
+          draggable: true,
+          progress: undefined,
+          theme: "dark",
+          });
+      }
+    }
+  };
+
```

Comment:
put all the required utility functions like formValidation to a separate file

Reply:
Move reducer functions to separate files